ffi/src/expressions.rs

```diff
@@ -245,3 +249,417 @@ pub extern "C" fn visit_expression_literal_bool(
+    /// Visit a 16bit `short`.
+    pub visit_short: extern "C" fn(data: *mut c_void, value: i16) -> usize,
+    /// Visit an 8bit `byte`.
+    pub visit_byte: extern "C" fn(data: *mut c_void, value: i8) -> usize,
+    /// Visit a 32bit `float`.
+    pub visit_float: extern "C" fn(data: *mut c_void, value: f32) -> usize,
+    /// Visit a 64bit `double`.
+    pub visit_double: extern "C" fn(data: *mut c_void, value: f64) -> usize,
+    /// Visit a `string`.
+    pub visit_string: extern "C" fn(data: *mut c_void, value: KernelStringSlice) -> usize,
+    /// Visit a `boolean`.
+    pub visit_bool: extern "C" fn(data: *mut c_void, value: bool) -> usize,
+    /// Visit a 64bit timestamp. The timestamp is microsecond precision and adjusted to UTC.
+    pub visit_timestamp: extern "C" fn(data: *mut c_void, value: i64) -> usize,
+    /// Visit a 64bit timestamp. The timestamp is microsecond precision with no timezone.
+    pub visit_timestamp_ntz: extern "C" fn(data: *mut c_void, value: i64) -> usize,
+    /// Visit a 32bit int date representing days since UNIX epoch 1970-01-01.
+    pub visit_date: extern "C" fn(data: *mut c_void, value: i32) -> usize,
+    /// Visit binary data at the `buffer` with length `len`.
+    pub visit_binary: extern "C" fn(data: *mut c_void, buffer: *const u8, len: usize) -> usize,
+    /// Visit a 128bit `decimal` value with the given precision and scale. The 128bit integer
+    /// is split into the most significant 64 bits in `value_ms`, and the least significant 64
+    /// bits in `value_ls`.
+    pub visit_decimal: extern "C" fn(
+        data: *mut c_void,
+        value_ms: u64, // Most significant 64 bits of decimal value
+        value_ls: u64, // Least significant 64 bits of decimal value
+        precision: u8,
+        scale: u8,
+    ) -> usize,
+    /// Visits a null value.
+    pub visit_null: extern "C" fn(data: *mut c_void) -> usize,
+    /// Visits an `and` expression which is made of a list of sub-expressions. This declares the
+    /// number of sub-expressions that the `and` expression will be made of. The visitor will populate
+    /// the list of expressions using the [`visit_variadic_sub_expr`] method.
+    pub visit_and: extern "C" fn(data: *mut c_void, len: usize) -> usize,
+    /// Visits an `or` expression which is made of a list of sub-expressions. This declares the
+    /// number of sub-expressions that the `or` expression will be made of. The visitor will populate
+    /// the list of expressions using the [`visit_variadic_sub_expr`] method.
+    pub visit_or: extern "C" fn(data: *mut c_void, len: usize) -> usize,
+    /// Visits a variadic sub-expression. This appends a sub-expression to a variadic expression
+    /// constructed in either [`visit_and`] or [`visit_or`]. The variadic expression is identified
+    /// by `variadic_id`, and the sub-expression is identified by `sub_expr_id`.
+    pub visit_variadic_sub_expr:
```

Comment:
Do we want to split up `visit_variadic_sub_expr` into `visit_or_sub_expr` and `visit_and_sub_expr`?